tactics/autorewrite.ml

```diff
@@ -278,5 +285,23 @@ let add_rew_rules ~locality base lrul =
   let open Goptions in
   match locality with
   | OptLocal -> cache_hintrewrite ((),(base,lrul))
-  | OptDefault | OptGlobal -> Lib.add_anonymous_leaf (inGlobalHintRewrite (base,lrul))
-  | OptExport -> Lib.add_anonymous_leaf (inExportHintRewrite (base,lrul))
+  | OptDefault ->
+    let () =
+      if not @@ Global.sections_are_opened () then
+        warn_deprecated_hint_rewrite_without_locality ()
+    in
+    Lib.add_anonymous_leaf (inGlobalHintRewrite (base,lrul))
```

Comment:
I can see that the current patch doesn't change the code flow and that this probably doesn't do anything more than the `OptLocal` case when inside a section. But suppose `inGlobalHintRewrite` was changed to support discharging. Then, the default behavior of `Hint Rewrite` inside a section would also change to be superglobal. So it seems to me that it would be safer and more readable to change this code to:

```ocaml
  | OptLocal | OptDefault when Global.sections_are_opened () -> cache_hintrewrite ((),(base,lrul))
  | OptDefault ->
    warn_deprecated_hint_rewrite_without_locality ();
    Lib.add_anonymous_leaf (inGlobalHintRewrite (base,lrul))
```

Don't you think?

Reply:
Yes, the current code is quite weird. I had to reread it three times to remember that despite the name `inGlobalHintRewrite` dies at section closure time. I already complained in the PR that introduced it that it was probably a bad idea to separate the original libobject in two different ones + a non-object. But as advocated below, `when` clauses easily introduce fragility. So I'd rather change the libobjects to follow the convention we have in other parts of the code, i.e. store a locality.